src/TSTransformer/nodes/expressions/transformBinaryExpression.ts

```diff
@@ -136,19 +140,42 @@ export function transformBinaryExpression(state: TransformState, node: ts.Binary
 		if (ts.isArrayLiteralExpression(node.left)) {
 			const rightExp = transformExpression(state, node.right);
 
+			// optimize empty array destructure
+			if (node.left.elements.length === 0) {
+				if (isUsedAsStatement(node) && luau.isArray(rightExp) && luau.list.isEmpty(rightExp.members)) {
+					return luau.none();
+				}
+				return rightExp;
+			}
+
 			if (luau.isCall(rightExp) && isLuaTupleType(state)(state.getType(node.right))) {
-				transformLuaTupleDestructure(state, node.left, rightExp);
+				transformOptimizedArrayAssignmentPattern(state, node.left, rightExp);
 				if (!isUsedAsStatement(node)) {
-					DiagnosticService.addDiagnostic(errors.noDestructureAssignmentExpression(node));
+					DiagnosticService.addDiagnostic(errors.noLuaTupleDestructureAssignmentExpression(node));
 				}
 				return luau.none();
 			}
 
+			if (luau.isArray(rightExp) && !luau.list.isEmpty(rightExp.members) && isUsedAsStatement(node)) {
+				transformOptimizedArrayAssignmentPattern(state, node.left, rightExp.members);
+				return luau.none();
+			}
+
 			const parentId = state.pushToVar(rightExp, "binding");
 			transformArrayAssignmentPattern(state, node.left, parentId);
 			return parentId;
 		} else if (ts.isObjectLiteralExpression(node.left)) {
-			const parentId = state.pushToVar(transformExpression(state, node.right), "binding");
+			const rightExp = transformExpression(state, node.right);
+
+			// optimize empty object destructure
+			if (node.left.properties.length === 0) {
+				if (isUsedAsStatement(node) && luau.isMap(rightExp) && luau.list.isEmpty(rightExp.fields)) {
+					return luau.none();
+				}
+				return rightExp;
```

Comment:
In the not luau.none() return case, this if statement will do nothing, because `wrapExpressionStatement` is still going to wrap that later on? Maybe if you do like a hacky `print(([] = [2]))` this will be slightly faster, but I don't think it's really a good thing to optimise. See comments above again.

Reply:
Yeah this just makes it so we get:
```Lua
local _ = {}
```
instead of
```Lua
local _binding = {}
```

The first is more clear that it's an expression statement. The second looks like broken code + unused variable.

Reply:
Oh, even does that with the luau.none() still? I thought it would be discarding the rightExp entirely

Reply:
Ahh no, my example was hypothetical. You'd have to have some array members. So more like:
```Lua
local _ = { 1, 2, 3 }
```
vs
```Lua
local _binding = { 1, 2, 3 }
```

No array members = erased

Reply:
This would be where `isUsedAsStatement()` passes, but `luau.isMap(rightExp) && luau.list.isEmpty(rightExp.fields)` fails

Reply:
Yeah, that's the case I think _especially_ should be kept in. Something that results in no array members will definitely be a bit finicky in the user code, so it's of greater emphasis to produce more sense-making emit. I think the `local _ = {}` case makes sense.